### ISSUE TITLE
Infinite loop when the token is expired

### ISSUE THREAD

**Issue by rekaburmeister** [aws-amplify/amplify-ui#1854]
### Before creating a new issue, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-ui/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-ui/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md#bug-reports).

### On which framework/platform are you having an issue?

React

### Which UI component?

Authenticator

### How is your app built?

Create React App

### Please describe your bug.

I have my app wrapped in the Authenticator as per the code snippet provided. This works perfectly fine until the token expires - at which point it seems to get itself into an infinite loop. `SignOut` gets called but the `localStorage` never gets cleared - potentially due to the process being interrupted by a redirect. 

I even tried completely removing my custom `SignIn` component and the `App` itself to see if it's my code's fault but the loop just kept on looping.

The only way to stop the loop is by clearing the local storage manually. Everything works fine afterwards until the token expires again (24h). I even added a `SignOut` button to see if that reproduces the issue (I have SSO set up with the signout flow disabled so the users aren't expected to sign out), but that works perfectly fine.

### What's the expected behaviour?

I'm not sure what's actually causing the loop. When trying to track down where the navigation is initiated from, execution gets to this method (in `Auth.ts`):

```typescript
	public async signOut(opts?: SignOutOpts): Promise<any> {
		try {
			await this.cleanCachedItems();
		} catch (e) {
			logger.debug('failed to clear cached items');
		}

		if (this.userPool) {
			const user = this.userPool.getCurrentUser();
			if (user) {
				await this.cognitoIdentitySignOut(opts, user);
			} else {
				logger.debug('no current Cognito user');
			}
		} else {
			logger.debug('no Cognito User pool');
		}

		/**
		 * Note for future refactor - no reliable way to get username with
		 * Cognito User Pools vs Identity when federating with Social Providers
		 * This is why we need a well structured session object that can be inspected
		 * and information passed back in the message below for Hub dispatch
		 */
		dispatchAuthEvent('signOut', this.user, `A user has been signed out`);
		this.user = null;
	}
```

Where `cleanCachedItems` is definitely called but it seems the items are never removed? The reload is then probably caused by `cognitoIdentitySignOut` being hit and a redirect being called.

### Help us reproduce the bug!

The only time I can reproduce this is when the token expires. 

My config is:

```
const aws_config = {
  aws_project_region: "eu-west-1",
  aws_cognito_region: "eu-west-1",
  aws_user_pools_id: "eu-west-1_myuserpoolid",
  aws_user_pools_web_client_id: "mylongclientid",
  oauth: {
    domain: "mydomain.eu-west-1.amazoncognito.com",
    scope: ["email", "openid", "aws.cognito.signin.user.admin"],
    // eslint-disable-next-line no-undef
    redirectSignIn: `${process.env.REACT_APP_REDIRECT_URL}`,
    // eslint-disable-next-line no-undef
    redirectSignOut: `${process.env.REACT_APP_REDIRECT_URL}`,
    responseType: "token",
  },
  federationTarget: "COGNITO_USER_POOLS",
};
export default aws_config;
```

I have another application using the same code and the same config, that works fine (though running with an older amplify version)

### Code Snippet

```typescript
...
import config from "./aws-config";

ReactDOM.render(
  <BrowserRouter>
    <React.StrictMode>
      <Authenticator
        hide={[SignIn, Greetings, RequireNewPassword]}
        amplifyConfig={config}
      >
        <MySignIn />
        <App />
      </Authenticator>
    </React.StrictMode>
  </BrowserRouter>,
  document.getElementById("root")
);

```

### Additional information and screenshots

```json
{
  "name": "myAp",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@carbon/grid": "^10.40.0",
    "@carbon/icons-react": "^10.45.0",
    "@carbon/type": "^10.40.0",
    "@testing-library/jest-dom": "^5.16.2",
    "@testing-library/react": "^12.1.2",
    "@testing-library/user-event": "^13.5.0",
    "@types/jest": "^27.4.0",
    "@types/react": "^17.0.39",
    "@types/react-dom": "^17.0.11",
    "@types/react-router-dom": "^5.3.3",
    "amazon-cognito-identity-js": "^5.2.8",
    "aws-amplify": "^4.3.19",
    "aws-amplify-react": "^5.1.9",
    "axios": "^0.25.0",
    "babel-polyfill": "^6.26.0",
    "carbon-components": "^10.52.0",
    "carbon-components-react": "^7.52.0",
    "carbon-icons": "^7.0.7",
    "classnames": "^2.3.1",
    "core-js": "^3.21.0",
    "env-cmd": "^10.1.0",
    "highlight.js": "^11.4.0",
    "jquery": "^3.6.0",
    "lodash.debounce": "^4.0.8",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-helmet-async": "^1.2.3",
    "react-router-dom": "^6.2.1",
    "react-scripts": "5.0.0",
    "use-dynamic-refs": "^1.0.0",
    "web-vitals": "^2.1.4"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "devDependencies": {
    "@types/carbon__icons-react": "^10.31.2",
    "@types/carbon-components-react": "^7.49.0",
    "@types/lodash.debounce": "^4.0.6",
    "@types/node": "^16.11.26",
    "@types/react-helmet": "^6.1.5",
    "@types/webpack": "^5.28.0",
    "@types/webpack-dev-server": "^4.7.2",
    "@typescript-eslint/eslint-plugin": "^5.10.2",
    "@typescript-eslint/parser": "^5.10.2",
    "css-loader": "^6.6.0",
    "css-minimizer-webpack-plugin": "^3.4.1",
    "eslint": "^8.8.0",
    "eslint-plugin-react": "^7.28.0",
    "html-webpack-plugin": "^5.5.0",
    "mini-css-extract-plugin": "^2.5.3",
    "postcss": "^8.4.6",
    "postcss-loader": "^6.2.1",
    "postcss-preset-env": "^7.3.1",
    "sass": "^1.49.7",
    "sass-loader": "^12.4.0",
    "style-loader": "^3.3.1",
    "ts-node": "^10.7.0",
    "typescript": "^4.6.3",
    "webpack": "^5.68.0"
  }
}

```

**Comment by wlee221** [comment]
Hi! It looks like you're on our legacy version of UI components. Can you try using `@aws-amplify/ui-react@2.x.x` instead of `aws-amplify-react-native`?

Here's our documentaiton:

https://ui.docs.amplify.aws/getting-started/installation
https://ui.docs.amplify.aws/components/authenticator

**Comment by rekaburmeister** [comment]
@wlee221 thanks for pointing this out to me, I didn't know there was a new version!! Looks much nicer, however, I have a set of issues/quetions:

- The documentation's first example (the [quick start](https://ui.docs.amplify.aws/components/authenticator#quick-start) ) has a line for `Amplify.configure`. `Amplify` is marked as deprecated for me: `@deprecated use named import`. Am I still using the wrong version or is this expected behaviour?
- As I stated in the original issue, we only provide SSO access for this particular application - so I think I need a completely custom Sign-in flow, right? I really like how you made the Sign-in page customizable but as far as I can see there is no option for federated Sign in (plus I already have my custom component ready). In the future, it would be super-cool if I could just pass in my custom component to use rather than customize parts of yours (or am I missing something here??)
- Your headless mode has the setup as wrapping `App` in `Authenticator.Provider`, in the export. I did this in `index.ts` - TypeScript was not happy with the example version. Unless I'm missing something here and these two implementations are not interchangeable, it might be worth including my version in your example as that will make TypeScript versions work as well.
- I'm assuming that it's ok to still access `Auth` directly? Such as `Auth.federatedSignIn(PROVIDER)` or `Auth.signOut()`

The original loop issue is now resolved, thank you!

**Comment by wlee221** [comment]
Glad it's resolved! Let me close this issue, but please feel free to open a new discussion or issue.

- Good catch, I think it's a miss on our documentation. The expected usage from js library is now

```
import { Amplify } from 'aws-amplify';
Amplify.configure(...);
```

we'll update the documentation.

- There is an option for federated sign in! Please see https://ui.docs.amplify.aws/components/authenticator#social-providers. Custom sign in component override is under our roadmap, but we don't have a specific date for that yet. But nevertheless, SSO should work! The only difference is that the new authenticator uses SSO with Cognito User Pool, while the legacy component used Cognito IDP only if I remember correctly. Let us know if you have trouble integrating it.
- `Authenticator.Provider` should not give any TS warnings. If it does, it's a bug -- please open a new issue 🙏  We test against TS apps, but we might have missed something.
- Sort of... 😄 We listen for `Auth.signOut`, and `federatedSignIn` should work as well because all it does it redirection. That said though, the happy case is if you use `Authenticator` and its underlying functions (`useAuthenticator`!) directly. But please feel free to open any new issues if you need any more `Auth` functionality!

This is a fairly new product, so again we gladly welcome any feedbacks and issues 🙏

**Comment by rekaburmeister** [comment]
I'm glad, too, been bugging me for a while! Plus, the new version looks really nice and is much more usable than the old one!

Regarding `{ Amplify }`, unless you super-quickly fixed it already, it was my fault for not noticing the brackets.

I did see the social providers, but I think what's confusing (me) here is that there is no obvious mention of one's own setup (even the [amplify](https://docs.amplify.aws/lib/auth/social/q/platform/js/#setup-your-auth-provider) docs seem like this in only for the 4 main providers). Reading the docs, my understanding is that I can use social providers here such as google or apple, not that I can use any federated identity provider that I set up in cognito (e.g. `foo`).  

So, e.g. if I have a SAML IDP set up as `foo` in cognito, would I be able to just do `<Authenticator socialProviders={['foo']}>` ?

Regarding the app wrapping, my bad again: the error I get is actually from eslint: `component definition is missing display name`

Finally, on `Auth.federatedSignIn()` - if I wanted to use `Authenticator` functions here, which one would that be? It seems there isn't an obvious one (`toFederatedSignIn` sounds more like it would take me to the page displaying my options of federated sign ins, rather than attempting to log in with them).

Let me know if you wanted me to open any other issues for the above and thanks a lot for all your help!
